src/Database/LSMTree/Internal/Run/Construction.hs

```diff
@@ -113,76 +119,147 @@ new (NumEntries nentries) npages = do
      RunAcc s
-  -> ST s ( Maybe (PageAcc, [Index.Chunk])
+  -> ST s ( Maybe RawPage
           , Maybe Index.Chunk
           , Bloom SerialisedKey
           , CompactIndex
           , NumEntries
           )
 unsafeFinalise racc@RunAcc {..} = do
-    mpage <- yield racc
-    (mchunk, index) <- Index.unsafeEnd mindex
+    mpagemchunk <- flushPageIfNonEmpty racc
+    (mchunk', index) <- Index.unsafeEnd mindex
     bloom <- Bloom.unsafeFreeze mbloom
-    numEntries <- NumEntries <$> readPrimVar entryCount
-    pure (mpage, mchunk, bloom, index, numEntries)
+    numEntries <- readPrimVar entryCount
+    let !mpage  = fst <$> mpagemchunk
+        !mchunk = selectChunk mpagemchunk mchunk'
+    pure (mpage, mchunk, bloom, index, NumEntries numEntries)
+  where
+    selectChunk :: Maybe (RawPage, Maybe Index.Chunk)
+                -> Maybe Index.Chunk
+                -> Maybe Index.Chunk
+    selectChunk (Just (_page, Just _chunk)) (Just _chunk') =
+        -- If flushing the page accumulator gives us an index chunk then
+        -- the index can't have any more chunks when we finalise the index.
+        error "unsafeFinalise: impossible double final chunk"
+    selectChunk (Just (_page, Just chunk)) _ = Just chunk
+    selectChunk _ (Just chunk)               = Just chunk
+    selectChunk _ _                          = Nothing
 
 -- | Add a serialised k\/op pair with an optional blob span. Use only for
 -- entries that are fully in-memory. Otherwise, use 'addChunkedKOp'.
 addFullKOp ::
      RunAcc s
   -> SerialisedKey
   -> Entry SerialisedValue BlobSpan
-  -> ST s (Maybe (PageAcc, [Index.Chunk]))
-addFullKOp racc@RunAcc{..} k e = do
-    modifyPrimVar entryCount (+ 1)
+  -> ST s ([RawPage], [RawOverflowPage], [Index.Chunk])
+addFullKOp racc k e
+  | PageAcc.entryWouldFitInPage k e = smallToLarge <$> addSmallKeyOp racc k e
+  | otherwise                       =                  addLargeKeyOp racc k e
+  where
+    smallToLarge :: Maybe (RawPage, Maybe Index.Chunk)
+                 -> ([RawPage], [RawOverflowPage], [Index.Chunk])
+    smallToLarge Nothing                   = ([],     [], [])
+    smallToLarge (Just (page, Nothing))    = ([page], [], [])
+    smallToLarge (Just (page, Just chunk)) = ([page], [], [chunk])
+
+addSmallKeyOp
+  :: RunAcc s
+  -> SerialisedKey
+  -> Entry SerialisedValue BlobSpan
+  -> ST s (Maybe (RawPage, Maybe Index.Chunk))
+addSmallKeyOp racc@RunAcc{..} k e =
+  assert (PageAcc.entryWouldFitInPage k e) $ do
+    modifyPrimVar entryCount (+1)
+    Bloom.insert mbloom k
+
+    -- We have to force a page boundary when the range finder bits change.
+    -- This is a constraint from the compact index. To do this we remember the
+    -- range finder bits of the previously added key and compare them to the
+    -- range finder bits of the current key.
+    rfbits <- readPrimVar rangeFinderCurVal             -- previous
+    let !rfbits' = keyTopBits16 rangeFinderPrecision k  -- current
+
+    noPageBoundaryNeeded <-
```

Comment:
Having negated boolean conditions (and resulting double negation) often makes it harder for me to follow what's going on. Maybe something like `elemAddedOnThisPage` and then only in the branch below conclude that this means we need to flush?